core/src/main/java/org/elasticsearch/search/aggregations/pipeline/moving/MovFunctionPipelineAggregator.java

```diff
@@ -0,0 +1,165 @@
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+public class MovFunctionPipelineAggregator extends PipelineAggregator {
+    private final DocValueFormat formatter;
+    private final GapPolicy gapPolicy;
+    private final int window;
+    private MovModel function;
+    private final Script script;
+
+    MovFunctionPipelineAggregator(String name, String[] bucketsPaths, DocValueFormat formatter, GapPolicy gapPolicy,
+                                         int window, MovModel function, Script script, Map<String, Object> metadata) {
+        super(name, bucketsPaths, metadata);
+        this.formatter = formatter;
+        this.gapPolicy = gapPolicy;
+        this.window = window;
+        this.function = function;
+        this.script = script;
+    }
+
+    /**
+     * Read from a stream.
+     */
+    public MovFunctionPipelineAggregator(StreamInput in) throws IOException {
+        super(in);
+        formatter = in.readNamedWriteable(DocValueFormat.class);
+        gapPolicy = GapPolicy.readFrom(in);
+        window = in.readVInt();
+        function = in.readOptionalNamedWriteable(MovModel.class);
+        script = in.readOptionalWriteable(Script::new);
+    }
+
+    @Override
+    public void doWriteTo(StreamOutput out) throws IOException {
+        out.writeNamedWriteable(formatter);
+        gapPolicy.writeTo(out);
+        out.writeVInt(window);
+        out.writeOptionalNamedWriteable(function);
+        out.writeOptionalWriteable(script);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return MovFunctionPipelineAggregationBuilder.NAME;
+    }
+
+    @Override
+    public InternalAggregation reduce(InternalAggregation aggregation, ReduceContext reduceContext) {
+        InternalMultiBucketAggregation<? extends InternalMultiBucketAggregation, ? extends InternalMultiBucketAggregation.InternalBucket>
+            histo = (InternalMultiBucketAggregation<? extends InternalMultiBucketAggregation, ? extends
+            InternalMultiBucketAggregation.InternalBucket>) aggregation;
+        List<? extends InternalMultiBucketAggregation.InternalBucket> buckets = histo.getBuckets();
+        HistogramFactory factory = (HistogramFactory) histo;
+
+        List<Bucket> newBuckets = new ArrayList<>();
+        EvictingQueue<Double> values = new EvictingQueue<>(this.window);
+
+        // For script, if it exists
+        ExecutableScript executableScript = null;
+
+        if (script != null) {
+            Map<String, Object> vars = new HashMap<>();
+            ExecutableScript.Factory scriptFactory = reduceContext.scriptService()
+                .compile(script, ExecutableScript.AGGS_CONTEXT);
+
+            if (script.getParams() != null) {
+                vars.putAll(script.getParams());
+            }
+            vars.put("values", values.toArray(new Double[values.size()]));
```

Comment:
This is a great case to use the new script contexts. Can you instead create a MovingFunctionScript class and have the execute method take `double[] values`?